pkg/provisioning/provision_test.go

```diff
@@ -0,0 +1,511 @@
+	pipelineService.EXPECT().RemoveConnector(ctx, pl1, "con1")
+	connService.EXPECT().Delete(ctx, "con1")
+	pipelineService.EXPECT().RemoveConnector(ctx, pl1, "con2")
+	connService.EXPECT().Delete(ctx, "con2")
+	connService.EXPECT().RemoveProcessor(ctx, "con2", "proc1con")
+	procService.EXPECT().Delete(ctx, "proc1con")
+	pipelineService.EXPECT().RemoveProcessor(ctx, pl1, "proc1")
+	procService.EXPECT().Delete(ctx, "proc1")
+
+	// create new pipeline
+	pipelineService.
+		EXPECT().
+		Create(ctx, "pipeline1", pl1config, pipeline.ProvisionTypeConfig).
+		Return(pl1, nil)
+
+	connService.EXPECT().Create(ctx, "con1", connector.TypeSource, cfg1, connector.ProvisionTypeConfig).Return(source, nil)
+	pipelineService.EXPECT().AddConnector(ctx, pl1, "con1")
+	connService.EXPECT().Create(ctx, "con2", connector.TypeDestination, cfg2, connector.ProvisionTypeConfig).Return(destination, nil)
+	pipelineService.EXPECT().AddConnector(ctx, pl1, "con2")
+
+	procService.EXPECT().Create(ctx, "proc1con", "js", procParentConn, procCfg, processor.ProvisionTypeConfig)
+	connService.EXPECT().AddProcessor(ctx, "con2", "proc1con")
+	procService.EXPECT().Create(ctx, "proc1", "js", procParentPipeline, procCfg, processor.ProvisionTypeConfig)
+	pipelineService.EXPECT().AddProcessor(ctx, pl1, "proc1")
+
+	pipelineService.EXPECT().Start(ctx, connService, procService, pl1)
+
+	service := NewService(db, logger, pipelineService, connService, procService)
+	err := service.ProvisionConfigFile(context.Background(), "./test/provision1-pipeline-with-processors.yml")
+	is.NoErr(err)
+}
+
+func TestProvision_ExistingPipelineWithChangedPlugin(t *testing.T) {
+	is := is.New(t)
+	logger := log.Nop()
+	db := &inmemory.DB{}
+	_, ctx, _ := db.NewTransaction(context.Background(), true)
+	ctrl := gomock.NewController(t)
+	pipelineService := mock.NewPipelineService(ctrl)
+	connService := mock.NewConnectorService(ctrl)
+	procService := mock.NewProcessorService(ctrl)
+	connBuilder := connmock.Builder{Ctrl: ctrl}
+
+	pl2config := pipeline.Config{Name: pipeline2.Name, Description: pipeline2.Description}
+	pl2 := &pipeline.Instance{
+		ID:            "pipeline2",
+		Config:        pl2config,
+		Status:        pipeline.StatusRunning,
+		ProvisionedBy: pipeline.ProvisionTypeConfig,
+	}
+	modifiedSourceConfig := connector.Config{
+		Name:       "con1",
+		Settings:   map[string]string{"key": "my/key"},
+		Plugin:     "builtin:s3",
+		PipelineID: "pipeline2",
+	}
+
+	modifiedSource := connBuilder.NewSourceMock("con1", modifiedSourceConfig)
+
+	// pipeline already exists
+	pipelineService.EXPECT().Get(ctx, "pipeline2").Return(pl2, nil)
+
+	// the old source connector has a different Plugin
+	connService.EXPECT().Get(ctx, "con1").Return(modifiedSource, nil)
+	connService.EXPECT().Get(ctx, "con2").AnyTimes()
+
+	service := NewService(db, logger, pipelineService, connService, procService)
+	err := service.ProvisionConfigFile(context.Background(), "./test/provision2-simple-pipeline.yml")
+	is.True(err != nil)
```

Comment:
As mentioned in the code - this should return no error, the pipeline needs to be provisioned but a warning should be logged.